### PR TITLE
GetPackageId

### PR DESCRIPTION
**Pull request by freddydk**
Add new function to get package id from publisher, name and id

### CODE REVIEW

File: NuGet/Get-BcNuGetPackageId.ps1 [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        [string] $publisher,
+        [Parameter(Mandatory=$true)]
+        [string] $name,
+        [Parameter(Mandatory=$true)]
+        [string] $id,
+        [Parameter(Mandatory=$false)]
+        [string] $version = ''
+    )
+
+    try { $id = ([GUID]::Parse($id)).Guid } catch { throw "Package id must be a valid GUID: $id" }
```
Comment (mazhelez) [inline NuGet/Get-BcNuGetPackageId.ps1:40]: "Package id"? Isn't `$id` the app ID?

File: NuGet/Get-BcNuGetPackageId.ps1 [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    $nname = [nuGetFeed]::Normalize($name)
+    $npublisher = [nuGetFeed]::Normalize($publisher)
+    if ($nname -eq '') { throw "Package name is invalid: '$name'" }
```
Comment (mazhelez) [inline NuGet/Get-BcNuGetPackageId.ps1:43]: Maybe?

```suggestion
    if ($nname -eq '') { throw "App name is invalid: '$name'" }

```

File: NuGet/Get-BcNuGetPackageId.ps1 [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    if ($npublisher -eq '') { throw "Package publisher is invalid: '$publisher'" }
```
Comment (mazhelez) [inline NuGet/Get-BcNuGetPackageId.ps1:44]:
```suggestion
    if ($npublisher -eq '') { throw "App publisher is invalid: '$publisher'" }

```

File: NuGet/Get-BcNuGetPackageId.ps1 [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            $nname = $nname.Substring(0, $nname.Length - ($packageId.Length - 99))
+        }
+        else {
+            throw "Package id is too long: $packageId, unable to shorten it"
+        }
+        $packageId = $packageIdTemplate.replace('{id}',$id).replace('{name}',$nname).replace('{publisher}',$npublisher).replace('{version}',$version)
+    }
+    return $packageId
+}
+Export-ModuleMember -Function Get-BcNuGetPackageId
```
Comment (mazhelez) [inline NuGet/Get-BcNuGetPackageId.ps1:58]: I think `Export-ModuleMember` doesn't do anything in a .ps1 file.

File: NuGet/Get-BcNuGetPackageId.ps1 [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    )
+
+    try { $id = ([GUID]::Parse($id)).Guid } catch { throw "Package id must be a valid GUID: $id" }
+    $nname = [nuGetFeed]::Normalize($name)
+    $npublisher = [nuGetFeed]::Normalize($publisher)
+    if ($nname -eq '') { throw "Package name is invalid: '$name'" }
+    if ($npublisher -eq '') { throw "Package publisher is invalid: '$publisher'" }
+    
+    $packageId = $packageIdTemplate.replace('{id}',$id).replace('{name}',$nname).replace('{publisher}',$npublisher).replace('{version}',$version)
+    if ($packageId.Length -ge 100) {
```
Comment (mazhelez) [inline NuGet/Get-BcNuGetPackageId.ps1:47]: Add a comment on why the shortening is needed. Maybe also include it in the doc above.

**Review by mazhelez — COMMENTED** [review]
Left some comments, I think one more iteration is needed.

File: NuGet/Get-BcNuGetPackageId.ps1 (hunk 4 above)
Comment (freddydk) [reply NuGet/Get-BcNuGetPackageId.ps1:60]: This is how we export functions in all BcContainerHelper .ps1 files

File: NuGet/Get-BcNuGetPackageId.ps1 (hunk 1 above)
Comment (freddydk) [reply NuGet/Get-BcNuGetPackageId.ps1]: You are right

**Comment by freddydk** [comment]
Made Id optional based on a comment from @kine on WhatsApp

**Comment by kine** [comment]
The Id must be optional too for e.g. Platform or Application dependencies.

**Comment by freddydk** [comment]
> The Id must be optional too for e.g. Platform or Application dependencies.

Yes, I added that to the PR in the comment above.

**Review by mazhelez — DISMISSED** [review]

**Review by mazhelez — DISMISSED** [review]

**Review by mazhelez — APPROVED** [review]

**Pull request merged**